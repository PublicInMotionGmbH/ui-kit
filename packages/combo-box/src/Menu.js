import React from 'react'
import PropTypes from 'prop-types'
import cls from 'classnames'

import { prefix } from '@talixo/shared'

const moduleName = prefix('combo-box')

/**
 * Dropdown menu component.
 *
 * @param {object} props
 * @param {*} [props.getItemProps]
 * @param {number} [props.highlightedIndex]
 * @param {*} [props.itemComponent]
 * @param {array} [props.items]
 * @param {boolean} [props.loading]
 * @param {string || number} [props.maxHeight]
 * @param {string} [props.overflow]
 * @param {boolean} [props.separated]
 * @param {*} [props.selectedItem]
 * @param {object} [props.style]
 * @returns {React.Element}
 */
const AutocompleteMenu = props => {
  const {
    getItemProps, highlightedIndex, items, itemComponent: ItemComponent,
    loading, maxHeight, overflow, selectedItem, separated, style
  } = props

  const maxHeightStyle = {
    overflowY: 'auto',
    maxHeight,
    ...style
  }

  const wrapperClsNames = cls(`${moduleName}__options`, {
    [`${moduleName}__options--separated`]: separated
  })

  const generateClsNames = (item, index) => (
    cls(`${moduleName}__item`, {
      [`${moduleName}__item--highlighted`]: highlightedIndex === index,
      [`${moduleName}__item--selected`]: selectedItem === item,
      [`${moduleName}__item--overflow-truncate`]: overflow === 'truncate',
      [`${moduleName}__item--overflow-break`]: overflow === 'break'
    })
  )

  if (loading) {
    return (
      <div className={wrapperClsNames} style={maxHeight ? maxHeightStyle : style}>
        <div
          className={`${moduleName}__item`}
        >
          Loading...
        </div>
      </div>
    )
  }

  return (
    <div className={wrapperClsNames} style={maxHeight ? maxHeightStyle : style}>
      {items.map((item, index) => (
        <div
          {...getItemProps({ item })}
          key={ItemComponent ? index : item}
          className={generateClsNames(item, index)}
        >
          {ItemComponent ? <ItemComponent item={item} /> : item}
        </div>
      ))}
    </div>
  )
}

AutocompleteMenu.propTypes = {
  /** Returns the props applied to menu item */
  getItemProps: PropTypes.func,

  /** The currently highlighted item */
  highlightedIndex: PropTypes.number,

  /** Optional item component */
  itemComponent: PropTypes.func,

  /** Items array */
  items: PropTypes.array,

  /** Loading state */
  loading: PropTypes.bool,

  /** Maximum toggle menu height */
  maxHeight: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),

  /** Item text overflow type */
  overflow: PropTypes.oneOf(['truncate', 'break']),

  /** The currently selected item */
  selectedItem: PropTypes.any,

  /** Displays menu as separated element */
  separated: PropTypes.bool,

  /** The currently selected item */
  style: PropTypes.object
}

AutocompleteMenu.defaultProps = {
  items: [],
  separated: false,
  getItemProps: props => props
}

export default AutocompleteMenu
