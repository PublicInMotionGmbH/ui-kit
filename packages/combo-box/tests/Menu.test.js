import React from 'react'
import { shallow } from 'enzyme'

import Menu from '../src/Menu'

describe('<Menu />', () => {
  it('should render correctly', () => {
    const wrapper = shallow(
      <Menu
        buildItemId={item => item}
        getItemProps={() => {}}
        renderItem={item => item}
      />
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('should render passed items', () => {
    const render = item => item
    const getItemProps = () => ({})

    const wrapper = shallow(
      <Menu
        options={[ 'abc', 'def', 'ghi' ]}
        buildItemId={item => item}
        getItemProps={getItemProps}
        renderItem={render}
      />
    )

    expect(wrapper.children().length).toBe(3)
    expect(wrapper.children().at(0).prop('item')).toBe('abc')
    expect(wrapper.children().at(1).prop('item')).toBe('def')
    expect(wrapper.children().at(2).prop('item')).toBe('ghi')
  })

  it('should pass correct props to items', () => {
    const render = item => item
    const getItemProps = () => ({})

    const wrapper = shallow(
      <Menu
        options={[ 'abc', 'def', 'ghi' ]}
        buildItemId={item => item}
        getItemProps={getItemProps}
        renderItem={render}
      />
    )

    expect(wrapper.children().at(0).prop('renderItem')).toBe(render)
    expect(wrapper.children().at(0).prop('getItemProps')).toBe(getItemProps)
  })

  it('should build keys for items', () => {
    const render = item => item
    const getItemProps = () => ({})

    const wrapper = shallow(
      <Menu
        options={[ 'abc', 'def', 'ghi' ]}
        buildItemId={(item, index) => index}
        getItemProps={getItemProps}
        renderItem={render}
      />
    )

    expect(wrapper.children().at(0).key()).toBe('0')
    expect(wrapper.children().at(1).key()).toBe('1')
    expect(wrapper.children().at(2).key()).toBe('2')

    wrapper.setProps({ buildItemId: item => item })

    expect(wrapper.children().at(0).key()).toBe('abc')
    expect(wrapper.children().at(1).key()).toBe('def')
    expect(wrapper.children().at(2).key()).toBe('ghi')
  })

  it('should render footer', () => {
    const render = item => item
    const getItemProps = () => ({})

    const wrapper = shallow(
      <Menu
        options={[ 'abc', 'def', 'ghi' ]}
        buildItemId={item => item}
        getItemProps={getItemProps}
        renderItem={render}
        footer={<div>FOOTER</div>}
      />
    )

    expect(wrapper.find('.talixo-combo-box__menu--footer').text()).toBe('FOOTER')
  })
})
