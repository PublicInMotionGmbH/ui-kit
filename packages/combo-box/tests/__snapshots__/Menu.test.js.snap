// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Menu /> renders correctly 1`] = `
<div
  className="talixo-combo-box__options"
>
  <div
    className="talixo-combo-box__item"
    item={1}
    key="1"
  >
    1
  </div>
  <div
    className="talixo-combo-box__item"
    item={3}
    key="3"
  >
    3
  </div>
  <div
    className="talixo-combo-box__item"
    item={5}
    key="5"
  >
    5
  </div>
</div>
`;

exports[`<Menu /> renders item component correctly 1`] = `
<div
  className="talixo-combo-box__options"
>
  <div
    className="talixo-combo-box__item talixo-combo-box__item--overflow-break"
    item={
      Object {
        "value": 1,
      }
    }
    key="0"
  >
    <div>
      <span>
        1
      </span>
    </div>
  </div>
  <div
    className="talixo-combo-box__item talixo-combo-box__item--overflow-break"
    item={
      Object {
        "value": 3,
      }
    }
    key="1"
  >
    <div>
      <span>
        3
      </span>
    </div>
  </div>
  <div
    className="talixo-combo-box__item talixo-combo-box__item--overflow-break"
    item={
      Object {
        "value": 5,
      }
    }
    key="2"
  >
    <div>
      <span>
        5
      </span>
    </div>
  </div>
</div>
`;
